### PR TITLE
In-memory Engine: reload manual loaded regions after leader transfer

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: ref https://github.com/tikv/tikv/issues/16966

What's Changed:

```commit-message
This PR addresses the problem where manually loaded ranges were not
automatically reloaded after a leader transfer.

To resolve this, a "manual-load-range" has been introduced, ensuring
that regions covered by this range are properly reloaded when they
handle write requests.
```

~I mark this as a draft PR, because it may increase the likelihood of a deadlock issue https://github.com/tikv/tikv/issues/17260 , as it could lead to a significant amount of eviction and loading.~
Edit: The deadlock issue has been resolved by #17331.

### Check List

Tests 

- Unit test
- Integration test

### Release note

```release-note
None
```

### CODE REVIEW

File: components/range_cache_memory_engine/src/engine.rs
Comment (glorv) [inline components/range_cache_memory_engine/src/engine.rs:410]: I prefer to handle this in ApplyObserver, but since other similar logic also handled here, I'm ok with this currently.

**Review by glorv — COMMENTED** [review]
Overall LGTM.

This PR does not handle range eviction but I think range eviction should avoid choosing preferred ranges as they will be loaded again when meet new write. Thus, hot range eviction or memroy high-water eviction should fist evict not perfered ranges.

File: components/range_cache_memory_engine/src/engine.rs
Comment (glorv) [inline components/range_cache_memory_engine/src/engine.rs:352]: We should avoid doing this kind of heavy check in this hot path.

I suggest check preferred_ranges in following condition:
1. When preferred range changes, trigges a batch load for all regions(leader) fall in this range
2. When region becomes leader, check and may load target region.

File: components/engine_traits/src/range_cache_engine.rs
Comment (glorv) [inline components/engine_traits/src/range_cache_engine.rs:167]: These two new methods looks weird to me under the current naming `CacheRegion`. Maybe we should split it into two struct like following:
```
struct Range {
  start: Vec<u8>,
  end: Vec<u8>,
}

struct CacheRegion {
  id: u64,
  epoch_version: u64,
  range: Range,
}
```
Thus all operations that actully take a range rather than a region should use `Range` instead, e.g, to use Range as the parameter to handle evict_by_range.

**Review by glorv — COMMENTED** [review]
BTW, Please also handle preferred ranges in region eviction. Since this PR will always trigger region load when new write comes, it should avoid preferred regions from being evicted to avoid load&evict loops.

File: components/engine_traits/src/range_cache_engine.rs
Comment (overvenus) [reply components/engine_traits/src/range_cache_engine.rs:171]: It's like a hundred-line refactor, I'll leave it to an other PR.

Comment (glorv) [reply components/engine_traits/src/range_cache_engine.rs:171]: Yeah, we may handle this later, won't block this pr.

File: components/hybrid_engine/src/observer/load.rs
Comment (glorv) [inline components/hybrid_engine/src/observer/load.rs:43]: Should only trigger load if this region range fall in any preferred ranges.

File: components/range_cache_memory_engine/src/range_manager.rs
Comment (glorv) [inline components/range_cache_memory_engine/src/range_manager.rs:266]: I don't see the benefit of using a `HashMap` here. Better replace it with `Vec` all the operations just iterate all the elements, or you can consider replace it with a `BTreeMap` like we do in `regions_by_range` as it provides better search performance.

File: components/engine_traits/src/range_cache_engine.rs
Comment (glorv) [inline components/engine_traits/src/range_cache_engine.rs:31]: This  naming is vague. Maybe better merge `Load` and `Eviction` to `RoleChange` and let IME to decide how to handle this role change. Thus, we can even avoiding adding the new `LoadObserver`

File: components/engine_traits/src/range_cache_engine.rs
Comment (overvenus) [reply components/engine_traits/src/range_cache_engine.rs:172]: We can do that, but I don't see the necessity.

File: components/range_cache_memory_engine/src/range_manager.rs
Comment (overvenus) [reply components/range_cache_memory_engine/src/range_manager.rs:266]: I'll use `Vec`. as impl `PartialOrd` and `Ord` for `CacheRegion` seems odd to me.

**Review by glorv — APPROVED** [review]

**Review by SpadeA-Tang — APPROVED** [review]

**Pull request merged**
